### ISSUE TITLE
Middleware Abort HTTP Request.

### ISSUE THREAD

**Issue by IppSec** [meteran/gnext#19]
Is it possible to have the middleware for Gnext to abort a request?  I'm trying to create protected endpoints that require authentication.  I have the endpoints hitting the authorization function itself but am having trouble finding a way to end the function within the middleware.

**Comment by meteran** [comment]
Hi @IppSec, thanks for your interest in gNext and question. The documention is still in progress, I have to find some more time to finish it. However, to not to stop you, I will write a longer answer to show how middlewares can be used.

General rule: if any handler or middleware returns an `error`, which is not `nil`, the whole execution flow is stopped and the error handler is being called. After the error handler is called, the execution is continued using After middlewares from the point it was stopped. Example:
```go
r := Router()
r.OnError(func(err error) (Status, interface{}) {
  return 422, nil
})

r.Use(Middleware{
  Before: Before1,
  After:  After1,
})
r.Use(Middleware{
  Before: Before2,
  After:  After2,
})
r.GET("/path", Handler)
```
The normal execution flow would be: `Before1->Before2->Handler->After2->After1`. However, if the handler `Before2` returns an error, the execution will look like: `Before1->Before2(error)->ErrorHandler->After2->After1`. If middleware `Before1` returns an error, execution will look like: `Before1(error)->ErrorHandler->After1`.

That's the theory. Going back to your example, you need to return an error from you authorization middleware and handler that error in error handler:
```go
type Headers struct {
  gnext.Headers
  Authorization    string `header:"Authorization"`
}

type AuthError struct { error }

func Authorize(headers *Headers) (*User, error) {
  if headers.Authorization == "" {
    return nil, &AuthError{fmt.Errorf("no token")}
  }
  // authorize
}

func ErrorHandler(err error) (gnext.Status, interface{}) {
  switch err.(type) {
  case *AuthError:
    return 401, nil
  default:
    return 500, nil
  }
}

func handler(usr *User) *User {
  fmt.Println("this should not be called in case of authorization error")
  return usr
}

func main() {
  r := gnext.Router()
  r.OnError(ErrorHandler)
  r.Use(gnext.Middleware{
    Before: Authorize,
  })
  r.GET("/path", handler)
  
  _ = r.Run()
}
```

This should do the work you need. Now with the gNext 0.7.0, you can also setup a dedicated error handler for a specific error:
```go
func ErrorHandler(err error) (gnext.Status, interface{}) {
  return 500, nil
}

func AuthErrorHandler(err *AuthError) (gnext.Status, interface{}) {
  return 401, nil
}

// usage:
r.OnError(ErrorHandler)
r.OnError(AuthErrorHandler)
```

In such implementation, the input parameter from error handler is recognized and if the specific error is raised, the error router will run the right handler. In the input is `error`, this will be a default error handler (in case there is no error handler for the current error type).

Also if an error handler returns a defined struct as response body, it will be added to the docs.

Hope it will help you. If you have any further questions, just ask :+1:

**Comment by IppSec** [comment]
Thank you so much for the detailed comment. This solved my issue perfectly, I'm new to golang and spent a lot of time looking at Gin Examples, which made it a lot more complicated than this.

The piece I loved about your example is this:
```  r.Use(gnext.Middleware{
    Before: Authorize,
  })```
  
I never thought to use the gnext.Middleware in the Use Statement. I spent a lot of time trying to create a function around gnext.Middleware, similar to how gin does it with Context.

Closing as this has completely solved my problem, already tested!
